Spring Cleaning: New Note Info Sub-task

a8ck3n opened the pull request:
### Fix

This PR adds the new navigation button and implements the new Note Info UI behind it.

![ezgif-7-c9243cb9d8ce](https://user-images.githubusercontent.com/40267301/82117025-34864e00-97a0-11ea-80a8-e3e343517345.gif)

Sub-task of #741.

### Test Button

1. Select any note
2. Confirm the new "..." button is visible
3. Confirm the Note Info view appears correctly

### Test View

1. Confirm table ignores taps
2. Confirm note details are correct
3. Confirm bottom sheet responds to swipes
4. Confirm "x" button dismisses view

### Review

This PR is to merge changes to the `741-note-options` "feature" branch. Not into `develop`.

**Notes for** @SylvesterWilmott 

1. The "x" button doesn't match the system default. May need a new image to get this exactly right. Currently uses an existing image from the project.
2. Date formats do not match design spec. Currently using the localized system settings (ie: the user's settings). The Note Options class has a date formatter that matches the spec too, if that is preferred.
3. I'll need to know some min/max sizes for the sheet on the iPad. Right now it does some rough sizing but it's not what it should be.

Please let me know if there is anything that looks off!

**Notes for** @jleandroperez 

- I've set up the `741-note-options` branch as the "feature" branch.
- I went with a XIB here as storyboards where a bit harder to wrangle into a bottom sheet like this. No layout in code.
- The UITableView uses standard UITableViewCell for now as these match the design spec. We can always introduce a custom subclass if needed though.
- The Note Info view is fully localized in code.
- I set up the "x" button to use a block-based callback instead of a delegate.
- Local data is declared as optional. Not sure what the best practice is for Swift and would love to hear your thoughts on this.

I'm thinking`var note: Note?` vs `var note: Note!` vs ? If we went with `!` we could test and fatal in `viewDidLoad` but assume the values are set everywhere else, much like how IBOutlets are handled.

### Release

Initial release notes are already part of the `741-note-options` branch. Those will need to be reviewed before merging the feature branch back to `develop` but should be fine for now.

SylvesterWilmott commented:
>The "x" button doesn't match the system default. May need a new image to get this exactly right. Currently uses an existing image from the project.

What you've used looks good at this point!

>Date formats do not match design spec. Currently using the localized system settings (ie: the user's settings). The Note Options class has a date formatter that matches the spec too, if that is preferred.

LGTM, thanks!

>I'll need to know some min/max sizes for the sheet on the iPad. Right now it does some rough sizing but it's not what it should be.

On iPad we should use a popover and we can let the system handle the width.

Few more notes:
- There appears to be an additional empty row at the bottom of the sheet. Instead, we should provide an empty area to respect safe zones and a 36pt empty space on devices where safe zones are not required.
- The "Information" title should use the `headline` type
- Text and icon colours should be consistent with other parts of the app
  - Icon/light text = gray50
  - Dark Text = gray100

a8ck3n commented:
Thanks for the feedback @SylvesterWilmott — very much appreciated!

I have some design changes here that you might want to review.

> The "Information" title should use the `headline` type

Done!

> The sheet in dark mode appears black. We should use the same colour as the rest of the UI using a blur effect on the sheet material to reinforce the overlay.

Also done! I've copied the styling of the existing "Action" sheet. I've also tweaked the display to handle any system provided safe areas as you noted. This is what we have so far:

![Dark-vs-Light-vs-Safe](https://user-images.githubusercontent.com/40267301/82343834-e71b1280-9a25-11ea-914b-a868853bee19.png)

The only thing I'm not sure about is the font colours. Is **gray50** the Dark mode font colour and **gray100** the colour for Light mode? The only example of a proper UITableView is the Settings which is what it currently mimics. I do see `UIColor.simplenoteTextColor` as an option though. With that the sheet looks like this:

![simplenoteTextColor](https://user-images.githubusercontent.com/40267301/82347397-20ee1800-9a2a-11ea-8a5d-3a7966d1ae82.png)

> On iPad we should use a popover and we can let the system handle the width.

I've done a first pass at this. If you'd like us to space it out a bit more, please let me know. Currently it looks like this:

![iPad-Dark-Popover](https://user-images.githubusercontent.com/40267301/82342625-6ad3ff80-9a24-11ea-8c4d-70eb1d9c25b0.png)

Again, thank you very much for the feedback. Let me know if anything else looks off!

jleandroperez commented on Simplenote/Classes/NoteInfoViewController.swift:
```diff
@@ -0,0 +1,157 @@
+        tablewView.tableFooterView = UIView()
+        
+        if hasSafeAreas {
+            self.view.frame.size.height -= 36
```
Tricky one! Ideally speaking, you'd always want to use Autolayout (and get rid of magic numbers, such as `-36`). In order to nuke this one, there are a couple things that could be done:

1. Pass along the parent's `safeAreaInsets` by means of the `additionalSafeAreaInsets`
  > infoViewController.additionalSafeAreaInsets = self.view.safeAreaInsets;

2. Adjust the layout here, accordingly to the `additionalSafeAreaInsets`

3. Since you're using `SPActionSheet`, you'd probably want to layout the main view, and adjust its size to the intrinsicContentSize. There are few (manual) ways to do this, involving querying the tableView's `contentSize`, and a bit or arithmetic. **But** main goal is always to get rid of inline constants

a8ck3n replied:
In this case it's a bit tricky as the 36 points are an arbitrary element of the design spec. I'm not sure how to account for them with Auto Layout. So what I've done is reversed the initial logic and removed that padding from the XIB, then only add the padding to the view if the container doesn't have safe area insets applied. The end result is one magic number (as a constant) in only one place instead of two.

Not sure this is the best approach but it's better. Would love to hear your thoughts on how to better tackle this!

jleandroperez replied:
Couple notes on this one!

- Yes, definitely a tricky problem
- This is actually similar to precalculating TableViewCell's height(s), once, for performance
- [Sample approach here](https://github.com/wordpress-mobile/WordPress-iOS/blob/d0564d3bc8f3d47fbb3e83e1017741f38329429c/WordPress/Classes/ViewRelated/Blog/Blog%20Details/BlogDetailsViewController.m#L343)
- And [this one might also help](https://stackoverflow.com/questions/18746929/using-auto-layout-in-uitableview-for-dynamic-cell-layouts-variable-row-heights)

In broad lines, solving this 100% with Autolayout would probably involve:

1. Defining everything with Autolayout Constraints, usage of [systemLayoutSizeFittingSize](https://developer.apple.com/documentation/uikit/uiview/1622624-systemlayoutsizefittingsize?language=objc) to determine the "minimum required size", 
2. Potentially using [this trick](https://github.com/wordpress-mobile/WordPress-iOS/blob/309e5249dce755650be4ccf8be9e818b237b51e4/WordPress/Classes/ViewRelated/Notifications/Tools/IntrinsicTableView.swift)
3. Forcing the layout system to calculate the minimum required size
4. And finally just doing the frame assignment

(4) is only required because the container itself isn't Autolayout friendly. But with an overhauled SPActionSheet none of this will probably be required, at all.

jleandroperez commented on Simplenote/Classes/SPNoteEditorViewController.m:
```diff
@@ -1982,8 +2032,66 @@ - (void)viewVersionAction:(id)sender {
+        NSArray *views = @[infoViewController.view];
+        SPActionSheet *newActionSheet = [SPActionSheet showActionSheetInView:self.navigationController.view withMessage:nil withContentViewArray:views withButtonTitleArray:nil delegate:self];
+        newActionSheet.swipeToDismiss = YES;
+        newActionSheet.contentView.layer.cornerRadius = 15;
```
Probably a great idea to move the `15` to a constant

jleandroperez commented on Simplenote/Classes/NoteInfoViewController.swift:
```diff
@@ -0,0 +1,157 @@
+        
+        if dismissAction == nil {
+            dismissButton.isHidden = true
+        }
```
Update: I've realized that since SPActionSheet is not using ViewController containment, `viewWillAppear` will never get called. Good to keep this one in `viewDidLoad` (we probably wanna drop SPActionSheet!)

a8ck3n replied:
Agreed on eventually removing SPActionSheet.

a8ck3n commented on Simplenote/Classes/NoteInfoViewController.swift:
```diff
@@ -0,0 +1,157 @@
+        }
+        
+        // Date Formatter
+        let dateFormatter = localizedDateFormatter()
```
Yes, this was definitely on my mind as well. Should have dropped a TODO in there so as not to overlook this. I'll get this fixed up though.

a8ck3n commented on Simplenote/Classes/NoteInfoViewController.swift:
```diff
@@ -0,0 +1,157 @@
+            dismissButton.isHidden = true
+        }
+        
+        tablewView.dataSource = self
```
Nice catch! Done!

a8ck3n commented on Simplenote/Classes/NoteInfoViewController.swift:
```diff
@@ -0,0 +1,157 @@
+    
+    @IBOutlet weak var titleLabel: UILabel!
+    @IBOutlet weak var dismissButton: UIButton!
+    @IBOutlet weak var tablewView: UITableView!
```
Thanks for letting me know. I saw this in your recent macOS changes and was thinking I should review best practices here.

a8ck3n commented on Simplenote/Classes/NoteInfoViewController.swift:
```diff
@@ -0,0 +1,157 @@
+    
+    func setUpTableCells() {
+        
+        cells.removeAll()
```
These cells are never refreshed, which is why I skipped setting a reuse identifier. They are set up once when first loading the view and then the same four static cells are used for the life of the view. As I understand it, the reuse identifier is largely for recycling cells during scrolling, which we don't do here.

This changes slightly with the addition of refresh on sync, but we still don't recycle cells. We just end up recreating the four static cells. The example I'd cite is when setting up a static table in a storyboard. Those cells do not have reuse identifiers assigned by default.

Let me know if you still think it makes sense to assign them. It won't hurt but I generally don't set up features unless I'm intentionally planning to use them. I feel it adds some clarity when revisiting the code.

jleandroperez replied:
I do agree that it's just a few cells, and performance shouldn't be hit in a noticeable way, in this particular scenario. 

But even if relying on the cell recycling mechanism is off the table, by the time a remote update comes thru, these UI components are already instantiated, and releasing / recreating them is not an ideal approach (plus non standard).

An alternative could be to just have a couple of private properties (referencing the cells), and updating the strings with an assignment.

a8ck3n commented on Simplenote/Classes/NoteInfoViewController.swift:
```diff
@@ -0,0 +1,157 @@
+    @IBOutlet weak var dismissButton: UIButton!
+    @IBOutlet weak var tablewView: UITableView!
+    
+    @objc var dismissAction: (() -> Void)? = nil
```
Thank you! I'll remove that assignment.

a8ck3n commented on Simplenote/Classes/NoteInfoViewController.swift:
```diff
@@ -0,0 +1,157 @@
+    @IBAction func dismissButtonTapped(_ sender: UIButton) {
+        
+        if dismissAction != nil {
+            dismissAction!()
```
Fixed!

a8ck3n replied:
I still haven't gotten into the habit of using `guard` but will keep an eye out for places where this makes sense.

a8ck3n commented on Simplenote/Classes/SPNoteEditorViewController.m:
```diff
@@ -1982,8 +2032,66 @@ - (void)viewVersionAction:(id)sender {
     versionActionSheet.tapToDismiss = NO;
     versionActionSheet.cancelButtonIndex = 0;
+    
+    [self disableRotation];
```
It looks like it was disabled there because the Versions UI doesn't update properly. The layout is mostly fine but the selection gets a bit funky.

If you meant for the Info UI, it works there. I **never** use my phone in landscape so it didn't even occur to me to **not** disable it! :) — I've enabled rotation for the Info UI now.

a8ck3n commented on Simplenote/Classes/SPNoteEditorViewController.m:
```diff
@@ -806,6 +829,18 @@ - (void)setKeyboardVisible:(BOOL)keyboardVisible
+- (void)disableRotation
+{
+    [(SPNavigationController *)self.navigationController setDisableRotation:YES];
+}
```
Yup, that makes sense. I've flagged it for removal once the new Versions UI is in place. 👍

a8ck3n commented on Simplenote/Classes/SPNoteEditorViewController.m:
```diff
@@ -1982,8 +2032,66 @@ - (void)viewVersionAction:(id)sender {
+        infoViewController.preferredContentSize = newFrame.size;
+        
+        [self presentViewController:infoViewController animated:YES completion:^{
+            self.noteInfoViewController = nil;
```
I've never seen this as an issue in practice thanks to how the view lifecycle is managed so I've never bothered with the weak/strong dance in this specific case. I ran some quick tests to verify the controller is properly released at the end of the block too, just to be safe.

Looking through the project, I'm seeing the completion block is never used anywhere. I find this pattern to be super convenient but if you think we should avoid it, let me know.

a8ck3n commented on Simplenote/Classes/NoteInfoViewController.swift:
```diff
@@ -0,0 +1,157 @@
+        dateFormatter.dateFormat = "MMM d, yyyy, HH:mm"
+        
+        return dateFormatter
+    }
```
Yes, this was in here in case Sylvester requested we match the date formatting from the design spec. I'll remove this now!

jleandroperez replied:
Since there's no secondary flow (guard let made sense if you needed to print / do something else), you can synthesize this like:

```
dismissAction?()
```

(No need for anything else!)

a8ck3n commented on Simplenote/Classes/NoteInfoViewController.swift:
```diff
@@ -0,0 +1,157 @@
+        guard let note = note else {
+            print("No note set for sheet.")
+            return nil
+        }
```
I tackled this by copying the pattern used for IBOutlets. Basically making the note var `!` and testing for it in `viewDidLoad`. This seems cleaner and results in less code but I'd love your opinion on this one.

a8ck3n commented on Simplenote/Classes/NoteInfoViewController.swift:
```diff
@@ -0,0 +1,157 @@
+        array.append((label: lables[0], detail: modifictionDate))
+        array.append((label: lables[1], detail: creationDate))
+        array.append((label: lables[2], detail: wordCount))
+        array.append((label: lables[3], detail: charCount))
```
Thanks for the tip on this! I had a look at how this was handled elsewhere in the project and moved these into a private enum type.

a8ck3n commented on Simplenote/Classes/SPNoteEditorViewController.m:
```diff
@@ -1982,8 +2032,66 @@ - (void)viewVersionAction:(id)sender {
+}
 
-    [(SPNavigationController *)self.navigationController setDisableRotation:YES];
+- (void)viewInfoAction
```
Done! Broke it down a bit and yes, much more readable!

a8ck3n commented:
Hello @jleandroperez !

I've pushed some new code to address the following:

- Enabled scrolling on the UITableView so that all rows are visible when using large dynamic text sizes.
- Refreshes the data in response to sync updates.
- Fixed alignment of header label on modern devices. Not sure how to make this work across devices though as the tables Auto Layout constraints change across devices. Maybe the solution is to disable Auto Layout for the label and add those constraints in code?

I've also replied to your feedback inline. I think I've addressed everything. A few notes:

- Using `.simplenoteTextColor` for the labels in the UITableView. This results in us having one color for all labels which is slightly different from the default behaviour of emphasizing the description and using a softer color for the value. If you have any thoughts on a good secondary color, please let me know!
- Changed the Note to be `Note!` and do a check in `viewDidLoad`. This is the same pattern used for IBOutlets. Curious as to your thoughts on this.
- Also would love some thoughts on [this](https://github.com/Automattic/simplenote-ios/pull/756#discussion_r427883859) and [this](https://github.com/Automattic/simplenote-ios/pull/756#discussion_r427987080).

Thanks again for the time and effort here!

jleandroperez commented on Simplenote/Classes/NoteInfoViewController.swift:
```diff
@@ -0,0 +1,162 @@
+    @IBOutlet private weak var tablewView: UITableView!
+    
+    @objc var dismissAction: (() -> Void)?
+    @objc var note: Note!
```
`NoteInfoViewController` has a strong dependency towards `Note` (that is, the VC shouldn't even exist if there's no note). 

Although this approach works as well (the app will crash if the `note` is not set), you can actually improve this by requiring the Note via initializer:

```
init(note: Note) {
    self.note = note
    super.init(....
}
```

### Why that'd be an improvement:
Imagine there's not one, but multiple strong dependencies. In order to figure out which public properties **must** be set, you could:

1. Read thru the code. **Problem:** Takes too long!
2. Read the class documentation. **Problem:** Nobody really documents that (including this class!). Plus where should it go, in the property itself? at the top of the class?
3. Read the designated initializer

(3) is a super natural way to spot the strong requirements. The SDK itself is pretty much full of such examples. For instance, [NSPersistentStoreCoordinator](https://developer.apple.com/documentation/coredata/nspersistentstorecoordinator) has a single public initializer, which requires a **NSManagedObjectModel**.

### Why it's different from the IBOutlets approach:
IBOutlets themselves are set by the `xib` / `storyboard` unarchiver. Even if we wanted to, there's no way to easily hint there's something missing.
